flink-compiler/src/main/java/org/apache/flink/compiler/dag/BinaryUnionNode.java

```diff
@@ -266,4 +268,44 @@ public void computeOutputEstimates(DataStatistics statistics) {
 		this.estimatedOutputSize = in1.estimatedOutputSize > 0 && in2.estimatedOutputSize > 0 ?
 			in1.estimatedOutputSize + in2.estimatedOutputSize : -1;
 	}
+
+	public static class UnionSemanticProperties implements SemanticProperties {
+
+		@Override
+		public FieldSet getTargetFields(int input, int sourceField) {
+			if (input != 0 && input != 1) {
+				throw new IndexOutOfBoundsException();
```

Comment:
How about returning an exception that explains that unions only support input to be 0 or 1. ?